### PR TITLE
`LangevinDynamicsMove` now accepts `constraint_tolerance` parameter

### LINKED ISSUE #608: Allow integrator tolerance to be specified for mcmc.LangevinDynamicsMove
Currently, [`mcmc.LangevinDynamicsMove`](https://github.com/choderalab/openmmtools/blob/main/openmmtools/mcmc.py#L1018-L1156) does not allow the integrator tolerance to be specified. We should add this as an optional argument to the initializer and enable it to be a public attribute that the user can change later.

### PR DESCRIPTION
**Pull request by ijpulidos**
## Description
`LangevinDynamicsMove` now accepts a `constraint_tolerance` parameter (defaults to the same value as the previously used `LangevinIntegrator`).

Resolves #608 

## Todos
- [x] Implement feature / fix bug
- [x] Add [tests](https://github.com/choderalab/openmmtools/tree/master/openmmtools/tests)
- [x] Update [documentation](https://github.com/choderalab/openmmtools/tree/master/docs) as needed
- [x] Update [changelog](https://github.com/choderalab/openmmtools/blob/master/docs/releasehistory.rst) to summarize changes in behavior, enhancements, and bugfixes implemented in this PR

## Status
- [ ] Ready to go

## Changelog
```
``LangevinDynamicsMove`` now allows ``constraint_tolerance`` parameter and public attribute, for specifying the fraction of the constrained distance within which constraints are maintained for the integrator (Refer to `Openmm's documentation <http://docs.openmm.org/latest/api-python/generated/openmm.openmm.LangevinMiddleIntegrator.html#openmm.openmm.LangevinMiddleIntegrator.setConstraintTolerance>_ for more information).
```

### CODE REVIEW

File: openmmtools/mcmc.py [hunk 1]
```diff
@@ -1143,17 +1150,21 @@ def __getstate__(self):
         serialization = super(LangevinDynamicsMove, self).__getstate__()
         serialization['timestep'] = self.timestep
         serialization['collision_rate'] = self.collision_rate
+        serialization['constraint_tolerance'] = self.constraint_tolerance
```
Comment (mikemhenry) [inline openmmtools/mcmc.py:1153]: Do we need to add a default key here so that way someone can deserialize an older file without getting a key error here?

**Review by mikemhenry — CHANGES_REQUESTED** [review]

**Comment by mikemhenry** [comment]
Also maybe we should add to the changelog entry a link to the openmm doc for that setting.

**Comment by zhang-ivy** [comment]
Looks great, thanks!! And I agree with Mike's suggestions.

**Review by zhang-ivy — APPROVED** [review]

File: openmmtools/mcmc.py (hunk 1 above)
Comment (ijpulidos) [reply openmmtools/mcmc.py:1153]: Yes, good catch! Thanks.

**Comment by mikemhenry** [comment]
Serialization will be important if you want to load an `nc` simulated with an older version of openmmtools but want to look at it after this version comes out, see https://github.com/choderalab/openmmtools/issues/612

**Comment by zhang-ivy** [comment]
Yes I definitely agree it would be useful, but I just meant I’ll leave it up to you guys to figure out what you want to do

**Review by mikemhenry — CHANGES_REQUESTED** [review]
Accidently dismised my review.

The only blocker on this is that we need to handle the case where `constraint_tolerance` isn't in the nc file.

**Comment by ijpulidos** [comment]
> The only blocker on this is that we need to handle the case where `constraint_tolerance` isn't in the nc file.

@mikemhenry I just checked and it isn't a problem, to my surprise. I was kind of expecting for it to be a problem, so it took me a while to realize why I think it isn't. As far as I can see, the flow of the object when deserializing is something like the following.

1. The sampler gets created from the `nc` files with the [`from_storage` method](https://github.com/choderalab/openmmtools/blob/main/openmmtools/multistate/multistatesampler.py#L264).
2. This method then creates the instance of the `sampler` with the sampler's [`_instantiate_sampler_from_reporter` method](https://github.com/choderalab/openmmtools/blob/7034e41eea1030185adffdf5da81f5899b2ad8e3/openmmtools/multistate/multistatesampler.py#L287)
3. The previous step generates/deserialize instances of the `mcmc_moves` using the reporter's [`read_mcmc_moves` method](https://github.com/choderalab/openmmtools/blob/7034e41eea1030185adffdf5da81f5899b2ad8e3/openmmtools/multistate/multistatesampler.py#L949)
4. Lastly, after calling the utility deserializing function, `deserialize`, this function just creates new instances of the `mcmc_moves` objects by calling their respective `__init__` via the `__new__` method [here](https://github.com/choderalab/openmmtools/blob/7034e41eea1030185adffdf5da81f5899b2ad8e3/openmmtools/utils.py#L676), since we are not overloading `__new__` in the MCMCmove objects. Resulting in using the attributes that are serialized, falling back to defaults for those who were not serialized.

I hope that's understandable. Let me know if I am making myself clear here.

The good news is that in #613 we are also indirectly testing this, since we are using legacy/older versions of serialized systems in the test. But I don't know if we want a specific test for this checking the match between serialized attributes vs. the ones that are falling back to default when deserializing an older simulation. Generally speaking, we don't really have much tests for resuming simulations, as far as I can tell.

**Comment by mikemhenry** [comment]
> The good news is that in https://github.com/choderalab/openmmtools/pull/613 we are also indirectly testing this, since we are using legacy/older versions of serialized systems in the test.

This is great since it will check for regressions, I want to then get 613 merged in first so we can have that check for this PR.

**Comment by zhang-ivy** [comment]
@mikemhenry @ijpulidos : I merged #613 into `main` and then merged `main` into this branch and then double checked that `TestSerializedMultiStateSampler` (introduced in #613) is passing successfully in the latest GHA workflows. Are you ok with me merging this now?

**Comment by mikemhenry** [comment]
Sounds good, we can have John review the release PR

**Comment by zhang-ivy** [comment]
@mikemhenry Can you "approve" this PR? merging is blocked because you requested changes

**Review by mikemhenry — APPROVED** [review]

**Pull request merged**